### PR TITLE
Autoformat confusable units

### PR DESCRIPTION
**Pull request by covracer**
I've seen errors crop up from using the different micro and mu characters. Follow matching recommendations on which character to prefer for micro, ohm, and angstrom. References:
* Section 22.2 Letterlike Symbols, subsection Unit Symbols, page 877 of [The Unicode Standard, Version 15.0
](https://www.unicode.org/versions/Unicode15.0.0/UnicodeStandard-15.0.pdf)
* Section 2.5 Duplicated Characters of [Unicode Technical Report 25](https://www.unicode.org/reports/tr25/)
* [SI brochure](https://www.bipm.org/documents/20126/41483022/SI-Brochure-9-EN.pdf)
* https://github.com/unicode-org/icu/blob/main/icu4c/source/data/unidata/confusables.txt

### CODE REVIEW

File: scripts/update_ambiguous_characters.py [hunk 1]
```diff
@@ -25,22 +25,28 @@ def get_mapping_data() -> dict:
...
-/// This file is auto-generated by scripts/update_ambiguous_characters.py.
+/// This file is auto-generated by `scripts/update_ambiguous_characters.py`.
 
 use once_cell::sync::Lazy;
 use rustc_hash::FxHashMap;
 
 /// Via: <https://github.com/hediet/vscode-unicode-data/blob/main/out/ambiguous.json>
 /// See: <https://github.com/microsoft/vscode/blob/095ddabc52b82498ee7f718a34f9dd11d59099a8/src/vs/base/common/strings.ts#L1094>
-pub(crate) static CONFUSABLES: Lazy<FxHashMap<u32, u8>> = Lazy::new(|| {
+pub(crate) static CONFUSABLES: Lazy<FxHashMap<u32, u32>> = Lazy::new(|| {
```
Comment (MichaReiser) [inline scripts/update_ambiguous_characters.py:35]: Nit: I wonder if we should instead change the type to `<char, char>`. Having the `char`s in place could make the table a bit more readable and avoids the `unwrap` calls.

**Review by MichaReiser — COMMENTED** [review]
Thanks for your contribution. 

I'm not very familiar with unicode confusables. So bare with me ;) 

Our current table is based on the VS code table and the standard seems to differentiate between mixed script and other confusables. I believe we currently only test for mixed-script confusables (similar to [unicode-security](https://github.com/unicode-rs/unicode-security/blob/45178d16579a39c61e016333c8c20dc52fb9db26/src/tables.rs#L4314)). Do you know if these added scripts are also mixed script confusables or general confusables?

**Comment by covracer** [comment]
Thanks for taking a look!

The VS Code list only contains "... characters (key) that are confusable with a basic ascii [0x20 to 0x7E] character (value)". https://github.com/hediet/vscode-unicode-data/blob/a2ee3e17acaad2957f910cf6c46f6ea2c6eae753/index.ts#L159

I wonder why they did that. Was it an easy starting point and they haven't been motivated to build it out further? Would a full table would be too slow in node.js? 🤷

I too am new to Unicode security, and am still trying to figure out how to apply domain name spoofing concepts to code linting and autoformatting. TR 39 suggested reading TR 36 first, where I found this:

> While compatibility normalization and mixed-script detection can handle the majority of spoofing cases, they do not handle single-script confusables. https://www.unicode.org/reports/tr36/#Single_Script_Spoofing

So it seems like normalization comes first in the process, before mixed script or single script handling. The substitutions I'm adding are normalizations.

A longer discussion of normalization from TR 36:
> Fortunately the design of IDN prevents a huge number of spoofing attacks. All conformant users of [[IDNA2003](https://www.unicode.org/reports/tr36/#IDNA2003)] are required to process domain names to convert what are called [compatibility-equivalent](http://www.unicode.org/glossary/#compatibility_equivalent) characters into a unique form using a process called compatibility normalization (NFKC)—for more information on this, see [[UAX15](https://www.unicode.org/reports/tr36/#UAX15)]. This processing eliminates most possibilities for visual spoofing by mapping away a large number of visually confusable characters and sequences.

File: crates/ruff/src/rules/ruff/rules/confusables.rs [hunk 2]
```diff
@@ -2116,5 +2116,8 @@ pub(crate) static CONFUSABLES: Lazy<FxHashMap<u32, u8>> = Lazy::new(|| {
         (1059, 89),
         (65283, 35),
         (65307, 59),
+        (0x212B, 0x00C5), // ANGSTROM SIGN → LATIN CAPITAL LETTER A WITH RING ABOVE
+        (0x2126, 0x03A9), // OHM SIGN → GREEK CAPITAL LETTER OMEGA
+        (0x00B5, 0x03BC), // MICRO SIGN → GREEK SMALL LETTER MU
```
Comment (charliermarsh) [inline crates/ruff/src/rules/ruff/rules/confusables.rs:2121]: Based on Section 2.5 in https://www.unicode.org/reports/tr25/, should this also include Kelvin sign?

File: scripts/update_ambiguous_characters.py (hunk 1 above)
Comment (charliermarsh) [reply scripts/update_ambiguous_characters.py:35]: @MichaReiser - That seems reasonable to me, but will it increase the size of this map?

File: crates/ruff/src/rules/ruff/rules/confusables.rs (hunk 2 above)
Comment (covracer) [reply crates/ruff/src/rules/ruff/rules/confusables.rs:2121]: The list from VS Code covers Kelvin with `(8490, 75)`.

**Comment by zanieb** [comment]
I believe we were concerned about increasing the size of the map here but with use of `match` it shouldn't matter anymore? @covracer would you be interested in updating with the latest changes on `main` and we can try to get this merged?

**Comment by charliermarsh** [comment]
Thanks, and sorry for the extensive delay here.

**Comment by charliermarsh** [comment]
Note to self: revisit these ecosystem changes today.

**Comment by charliermarsh** [comment]
\cc @zanieb - Do you have an opinion on whether this should be a preview change...?

**Comment by zanieb** [comment]
@charliermarsh looks like it creates new violations and it's not a bug fix so it should probably be preview?

**Comment by charliermarsh** [comment]
Done in https://github.com/astral-sh/ruff/pull/8473.

**Comment by covracer** [comment]
Thanks! I've looked over the hits and made a best guess at how to proceed. Over time I'll try to test and submit Pull Requests.

> [examples/interaction/tools/subcoordinates_zoom.py:22:23:](https://github.com/bokeh/bokeh/blob/644813cd2e149f53ebfefc2e68ef60d421608ab4/examples/interaction/tools/subcoordinates_zoom.py#L22) RUF001 String ...

I would suggest applying the fix for this abbreviation of microvolts in [tooltip text](https://docs.bokeh.org/en/latest/docs/examples/interaction/tools/subcoordinates_zoom.html).

> [pandas/_libs/tslibs/timedeltas.pyi:58:6:](https://github.com/pandas-dev/pandas/blob/0cdb37c445fc0a0a2144f981fd1acbf4f07d88ee/pandas/_libs/tslibs/timedeltas.pyi#L58) RUF001 String ...
> [pandas/tests/scalar/timedelta/test_constructors.py:306:25:](https://github.com/pandas-dev/pandas/blob/0cdb37c445fc0a0a2144f981fd1acbf4f07d88ee/pandas/tests/scalar/timedelta/test_constructors.py#L306) RUF001 String ...

timedeltas.pyi and timedeltas.pyx list many aliases for microseconds (among other units). test_constructors.py tests using the aliases. I would suggest `# noqa: RUF001`.

> [pandas/core/indexes/base.py:5352:26:](https://github.com/pandas-dev/pandas/blob/0cdb37c445fc0a0a2144f981fd1acbf4f07d88ee/pandas/core/indexes/base.py#L5352) RUF003 Comment ...
> [pandas/core/indexes/base.py:5353:46:](https://github.com/pandas-dev/pandas/blob/0cdb37c445fc0a0a2144f981fd1acbf4f07d88ee/pandas/core/indexes/base.py#L5353) RUF003 Comment ...
> [pandas/core/indexes/base.py:5354:69:](https://github.com/pandas-dev/pandas/blob/0cdb37c445fc0a0a2144f981fd1acbf4f07d88ee/pandas/core/indexes/base.py#L5354) RUF003 Comment ...

This multiline comment documents performance analysis timings. I would suggest applying the fix.

> [pandas/tests/io/test_clipboard.py:67:34:](https://github.com/pandas-dev/pandas/blob/0cdb37c445fc0a0a2144f981fd1acbf4f07d88ee/pandas/tests/io/test_clipboard.py#L67) RUF001 String ...

This is part of `dict[str, list[str]]` test data for `utf8`. ASCII and not-confusable Latin-1 Supplement characters are included alongside characters from higher Unicode blocks. I would suggest applying the fix.

> [zerver/tests/test_invite.py:1281:32:](https://github.com/zulip/zulip/blob/0b3f7a5a6bc0736ec8eb209fe951f487969e86a6/zerver/tests/test_invite.py#L1281) RUF001 String ...
> [zerver/tests/test_signup.py:1021:29:](https://github.com/zulip/zulip/blob/0b3f7a5a6bc0736ec8eb209fe951f487969e86a6/zerver/tests/test_signup.py#L1021) RUF001 String ...

These tests use the same copy+pasted string to test "non-ASCII characters", 3 from the Basic Latin block and 6 from the Latin-1 Supplement block (5 of which aren't flagged as confusable). Unless the project would like to begin testing characters from higher blocks, I would suggest dropping the character or adding `# noqa: RUF001`.

**Comment by covracer** [comment]
Bokeh updated in https://github.com/bokeh/bokeh/pull/13668

**Comment by charliermarsh** [comment]
Thanks @covracer!

**Pull request merged**